core/includes/form.inc

```diff
@@ -1493,6 +1488,28 @@ function _form_validate(&$elements, &$form_state, $form_id = NULL) {
         call_user_func_array($function, array(&$elements, &$form_state, $form_state['complete_form']));
       }
     }
+
+    // Ensure that a #required form error is thrown, regardless of whether
+    // #element_validate handlers changed any properties. If $is_empty_value
+    // is defined, then above #required validation code ran, so the other
+    // variables are also known to be defined and we can test them again.
+    if (isset($is_empty_value) && ($is_empty_multiple || $is_empty_string || $is_empty_value || $is_empty_null)) {
```

Comment:
I'm not sure I understand why we need to duplicate these checks instead of using the `$elements['#required_but_empty']` status that was set in the above conditional.

Or alternatively, if we want to ensure this fires regardless of the `$elements['#required_but_empty']` value, perhaps we should set a single variable to check, instead of aggregating all these variables together again.